### PR TITLE
COSI-35: Improve Logging Clarity and Accuracy for Object Storage and IAM Operations

### PR DESCRIPTION
**Pull request by anurag4DSB**
This PR improves the logging structure and clarity across the object storage (S3) and IAM client operations, ensuring that logs are consistent, accurate, and useful for debugging, particularly in concurrent scenarios. Best practices inspired from [Kubernetes community guidelines](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-instrumentation/logging.md#what-method-to-use)

### Key Improvements:
	1.	Accurate Log Messages
	- Updated misleading or redundant log messages to reflect the exact operation being performed.
	- Example:
	       •	“Fetching S3 parameters” → “Extracting object storage parameters” (covers both IAM and S3).
	       •	“Fetching secret info” → “Validating object storage provider secret parameters” for improved clarity.
	2.	Self-Contained Logs for Debugging
	- Included identifying fields (e.g., secretName, namespace, userName, bucketName) in all log messages, including success, error, and skipped scenarios.
	- This makes logs easier to trace when multiple operations or secrets are being handled concurrently.
	3.	Centralized Success Logging
	- Higher-level methods (e.g., CreateBucketAccess, RevokeBucketAccess) now log successes only after completing all individual operations.
	- Lower-level methods no longer log redundant success messages, reducing noise.
	4.	Graceful Handling of Expected Errors
	- Improved clarity for cases where objects (like inline policies or access keys) do not exist:
	- Logs explicitly state that deletion is skipped instead of implying false success.
	- Example:
	       •	“Access key does not exist, skipping deletion” instead of “Successfully deleted access key.”
	5.	Improved Debug Verbosity
	- Refined logging verbosity levels:
	- V(2): Operational logs for meaningful events (e.g., creating a user, deleting a policy).
	- V(4): Detailed debug logs (e.g., TLS configuration, IAM endpoint overrides).

### Why These Changes Matter:
	•	Clearer Logs: Logs now provide accurate and detailed context, improving traceability and reducing ambiguity.
	•	Concurrency-Safe: Self-contained logs ensure clarity when multiple operations run concurrently.
	•	Reduced Redundancy: Lower-level methods focus on propagating errors, leaving centralized logging to higher-level methods.
	•	Debugging Efficiency: The use of verbosity levels ensures the right amount of detail is logged for normal operations (V(2)) and debugging (V(4) or V(5)).

### Testing:
	•	Verified logs are correctly emitted for various scenarios:
	•	Successful operations
	•	Missing or invalid parameters
	•	Skipped operations (e.g., non-existent policies or access keys)
	•	Concurrent executions (no interleaving ambiguity)

### CODE REVIEW

**Comment by anurag4DSB** [comment]
I thought of adding a helper log method like the one below, but it reduces readability.
Related to [this comment](https://github.com/scality/cosi-driver/pull/61#discussion_r1883176870) from @jonathan-gramain 

```go
// logOperation logs a structured log message with a consistent format.
// operation: The action being performed (e.g., "Initializing", "Fetching", "Validating").
// entity: The object/entity being operated on (e.g., "S3 client", "secret", "bucket").
// fields: Additional key-value pairs for context.
// success: Whether the operation succeeded or failed.
// err: (optional) The error if the operation failed).
// logLevel: Desired verbosity level (e.g., 2, 3, 4, etc.).
func logOperation(logLevel int, operation, entity string, fields map[string]interface{}, success bool, err error) {
	baseFields := []interface{}{
		"operation", operation,
		"entity", entity,
	}

	for k, v := range fields {
		baseFields = append(baseFields, k, v)
	}

	message := fmt.Sprintf("%s %s", operation, entity)

	if err != nil {
		klog.ErrorS(err, fmt.Sprintf("%s failed", message), baseFields...)
	} else if success {
		klog.V(logLevel).InfoS(fmt.Sprintf("%s succeeded", message), baseFields...)
	} else {
		klog.V(logLevel).InfoS(fmt.Sprintf("%s in progress", message), baseFields...)
	}
}
```

Example usage:
```
logOperation(4, "Fetching", "object storage provider secret", map[string]interface{}{"secretName": ospSecretName, "namespace": namespace}, true, nil)
```

File: pkg/clients/iam/iam_client.go [hunk 1]
```diff
@@ -110,12 +106,12 @@ func (client *IAMClient) AttachS3WildcardInlinePolicy(ctx context.Context, userN
 	}
 
 	_, err := client.IAMService.PutUserPolicy(ctx, input)
-	if err != nil {
-		return fmt.Errorf("failed to attach inline policy to IAM user %s: %w", userName, err)
-	}
+	// if err != nil {
```
Comment (BourgoisMickael) [inline pkg/clients/iam/iam_client.go:109]: Can you remove the comment completely ?

File: pkg/driver/provisioner_server_impl.go [hunk 2]
```diff
@@ -125,126 +123,31 @@ func (s *ProvisionerServer) DriverCreateBucket(ctx context.Context,
 		return nil, status.Error(codes.InvalidArgument, "unsupported client type for bucket creation")
 	}
 
+	klog.V(3).InfoS("Creating bucket", "bucketName", bucketName)
```
Comment (BourgoisMickael) [inline pkg/driver/provisioner_server_impl.go:126]: Can you use a constant or enum somewhere for log level so we know what 2, 3, 5 relates to

**Review by BourgoisMickael — APPROVED** [review]

File: pkg/clients/iam/iam_client.go (hunk 1 above)
Comment (anurag4DSB) [reply pkg/clients/iam/iam_client.go:109]: Oh yes, I will do that, its a left over.

File: pkg/driver/provisioner_server_impl.go (hunk 2 above)
Comment (anurag4DSB) [reply pkg/driver/provisioner_server_impl.go:126]: Actually its a standard practice in kubernetes and klog.
[Kubernetes community guidelines](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-instrumentation/logging.md#what-method-to-use)

Comment (anurag4DSB) [reply pkg/driver/provisioner_server_impl.go:126]: What do you think of 
```
const (
	LogLevelError = 0 // Error and critical logs
	LogLevelWarn  = 1 // Warnings
	LogLevelInfo  = 2 // General info messages
	LogLevelDebug = 3 // Debug-level messages
	LogLevelTrace = 5 // Detailed trace logs
)
```

Comment (anurag4DSB) [reply pkg/driver/provisioner_server_impl.go]: Added in the latest review commit: https://github.com/scality/cosi-driver/pull/67/commits/b2dfa498fffeb21d18c2af76e6696ccc9e6e810d

File: pkg/constants/constants.go [hunk 3]
```diff
@@ -0,0 +1,12 @@
+package constants
+
+// Log level constants for structured logging, starting from 1
+// 0 is default if no level is provided
+// Guidelines: https://github.com/kubernetes/community/blob/master/contributors/devel/sig-instrumentation/logging.md#what-method-to-use
+const (
+	LvlDefault = iota + 1 // 1 - General configuration, routine logs
```
Comment (BourgoisMickael) [inline pkg/constants/constants.go:7]: Why the different prefix `Lvl` vs `Log`.

Idk how a go project is organized, but would it be better organized if you keep that in a more specific file `loglevel.go` then you can name the consts without prefix ?

**Review by BourgoisMickael — APPROVED** [review]

File: pkg/constants/constants.go (hunk 3 above)
Comment (anurag4DSB) [reply pkg/constants/constants.go:7]: You are right we should keep it consistent.
I will change it to log.

From GO point of view, the package matters and not the file name.

Comment (anurag4DSB) [reply pkg/constants/constants.go:7]: I changed it, thanks for the catch.

**Review by BourgoisMickael — APPROVED** [review]

File: helm/scality-cosi-driver/values.yaml [hunk 4]
```diff
@@ -8,8 +8,8 @@ replicaCount: 1
 
 
 logLevels:
-  driver: "5"
-  sidecar: "5"
+  driver: "2"
```
Comment (tmacro) [inline helm/scality-cosi-driver/values.yaml:11]: Could be nice to add comment with the text name of the log level.

**Review by tmacro — APPROVED** [review]

**Comment by anurag4DSB** [comment]
Added 2 extra commits.
one for adding comments

And the last one for logging req in trace logs.

**Comment by anurag4DSB** [comment]
Rebased to squash review commits and pushed.

File: helm/scality-cosi-driver/values.yaml (hunk 4 above)
Comment (anurag4DSB) [reply helm/scality-cosi-driver/values.yaml:17]: Thanks added in kustomize files and helm

**Pull request merged**